src/dune/build_system.ml

```diff
@@ -1360,59 +1409,78 @@ end = struct
-    let prev_trace = Trace_db.get (Path.build head_target) in
+    let env = Internal_rule.effective_env rule in
+    let rule_loc = rule_loc ~file_tree:t.file_tree ~info ~dir in
     let sandbox_mode =
       match Action.is_useful_to_sandbox action with
       | Clearly_not ->
         let config = Dep.Set.sandbox_config deps in
         if Sandbox_config.mem config Sandbox_mode.none then
           Sandbox_mode.none
         else
-          User_error.raise
-            ~loc:(rule_loc ~file_tree:t.file_tree ~info ~dir)
+          User_error.raise ~loc:rule_loc
             [ Pp.text
               "Rule dependencies are configured to require sandboxing, but \
                the rule has no actions that could potentially require \
                sandboxing."
             ]
       | Maybe ->
-        select_sandbox_mode
-          ~loc:(rule_loc ~file_tree:t.file_tree ~info ~dir)
+        select_sandbox_mode ~loc:rule_loc
           (Dep.Set.sandbox_config deps)
           ~sandboxing_preference:t.sandboxing_preference
     in
-    let action_ctx = Action_exec.Context.make ~targets ~context ~env in
-    let rule_digest =
-      let env = Action_exec.Context.env action_ctx in
-      let trace =
-        ( Dep.Set.trace deps ~sandbox_mode ~env ~eval_pred
-        , List.map targets_as_list ~f:(fun p -> Path.to_string (Path.build p))
-        , Option.map context ~f:(fun c -> c.name)
-        , Action.for_shell action )
+    let* rule_need_rerun =
+      let force =
+        !Clflags.force
+        && List.exists targets_as_list ~f:Path.Build.is_alias_stamp_file
       in
-      Digest.generic trace
+      let always_rebuild = Dep.Set.has_universe deps in
+      if force || always_rebuild then
+        Fiber.return true
+      else
+        (* [prev_trace] will be [None] if rule is run for the first time. *)
+        let prev_trace = Trace_db.get (Path.build head_target) in
+        (* [targets_digest] will be [None] if not all targets were build. *)
+        let targets_digest = compute_targets_digest targets_as_list in
+        let rule_digest =
+          compute_rule_digest rule ~deps ~action ~sandbox_mode
+        in
+        let rule_or_targets_changed =
+          match (prev_trace, targets_digest) with
+          | Some prev_trace, Some targets_digest ->
+            prev_trace.rule_digest <> rule_digest
+            || prev_trace.targets_digest <> targets_digest
+          | _ -> true
+        in
+        if rule_or_targets_changed then
+          Fiber.return true
+        else
+          let prev_trace = Option.value_exn prev_trace in
+          let rec loop stages =
+            match stages with
+            | [] -> Fiber.return false
+            | (deps, old_digest) :: rest ->
+              let* () = build_deps deps in
+              let new_digest =
+                compute_dependencies_digest deps ~sandbox_mode ~env ~eval_pred
+              in
+              if old_digest <> new_digest then
+                Fiber.return false
```

Comment:
This is supposed to be `true`, I guess? I guess we need a test case that covers this.

Reply:
Ah, that's probably because I started looking at a random commit. Sorry, good to know this is caught.